perception/autoware_tensorrt_bevdet/src/bevdet.cpp

```diff
@@ -0,0 +1,717 @@
+      }
+    }
+  }
+
+  std::vector<int> kept;
+  for (int i = 0; i < num_points; i++) {
+    if (
+      static_cast<int>((frustum + i)->x()) >= 0 &&
+      static_cast<int>((frustum + i)->x()) < xgrid_num_ &&
+      static_cast<int>((frustum + i)->y()) >= 0 &&
+      static_cast<int>((frustum + i)->y()) < ygrid_num_ &&
+      static_cast<int>((frustum + i)->z()) >= 0 &&
+      static_cast<int>((frustum + i)->z()) < zgrid_num_) {
+      kept.push_back(i);
+    }
+  }
+
+  valid_feat_num_ = kept.size();
+  int * ranks_depth_host = new int[valid_feat_num_];
+  int * ranks_feat_host = new int[valid_feat_num_];
+  int * ranks_bev_host = new int[valid_feat_num_];
+  int * order = new int[valid_feat_num_];
+
+  for (int i = 0; i < valid_feat_num_; i++) {
+    Eigen::Vector3f & p = frustum[kept[i]];
+    ranks_bev_host[i] = static_cast<int>(p.z()) * xgrid_num_ * ygrid_num_ +
+                        static_cast<int>(p.y()) * xgrid_num_ + static_cast<int>(p.x());
+    order[i] = i;
+  }
+
+  thrust::sort_by_key(ranks_bev_host, ranks_bev_host + valid_feat_num_, order);
+  for (int i = 0; i < valid_feat_num_; i++) {
+    ranks_depth_host[i] = _ranks_depth[kept[order[i]]];
+    ranks_feat_host[i] = _ranks_feat[kept[order[i]]];
+  }
+
+  delete[] _ranks_depth;
+  delete[] _ranks_feat;
+  delete[] frustum;
+  delete[] order;
+
+  std::vector<int> interval_starts_host;
+  std::vector<int> interval_lengths_host;
+
+  interval_starts_host.push_back(0);
+  int len = 1;
+  for (int i = 1; i < valid_feat_num_; i++) {
+    if (ranks_bev_host[i] != ranks_bev_host[i - 1]) {
+      interval_starts_host.push_back(i);
+      interval_lengths_host.push_back(len);
+      len = 1;
+    } else {
+      len++;
+    }
+  }
+
+  interval_lengths_host.push_back(len);
+  unique_bev_num_ = interval_lengths_host.size();
+
+  int * interval_starts_host_ptr = new int[interval_starts_host.size()];
+  int * interval_lengths_host_ptr = new int[interval_lengths_host.size()];
+
+  memcpy(
+    interval_starts_host_ptr, interval_starts_host.data(),
+    interval_starts_host.size() * sizeof(int));
+  memcpy(
+    interval_lengths_host_ptr, interval_lengths_host.data(),
+    interval_lengths_host.size() * sizeof(int));
+
+  ranks_bev_ptr.reset(ranks_bev_host);
+  ranks_depth_ptr.reset(ranks_depth_host);
+  ranks_feat_ptr.reset(ranks_feat_host);
+  interval_starts_ptr.reset(interval_starts_host_ptr);
+  interval_lengths_ptr.reset(interval_lengths_host_ptr);
+
+  RCLCPP_INFO(rclcpp::get_logger("BEVDet"), "valid_feat_num: %d", valid_feat_num_);
+  RCLCPP_INFO(rclcpp::get_logger("BEVDet"), "unique_bev_num: %d", unique_bev_num_);
+}
+
+void print_dim(nvinfer1::Dims dim, const std::string & name)
+{
+  std::ostringstream oss;
+  oss << name << " : ";
+  for (auto i = 0; i < dim.nbDims; i++) {
+    oss << dim.d[i] << ' ';
+  }
+  RCLCPP_INFO(rclcpp::get_logger("BEVDet"), "%s", oss.str().c_str());
+}
+
+int BEVDet::initEngine(const std::string & engine_file)
```

Comment:
Namespaces:  Wrap the entire `BEVDet` class definition within a namespace to avoid potential naming conflicts.  For example: `namespace autoware_tensorrt_bevdet { class BEVDet { ... }; }`